cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -267,15 +267,68 @@ struct TransduceToken {
-    const bool is_part_of_invalid_line =
+    const bool is_part_of_valid_line =
       (match_id != static_cast<SymbolGroupT>(dfa_symbol_group_id::ERROR) &&
        state_id == static_cast<StateT>(TT_VLD));
 
     // Indicates whether we transition from an invalid line to a potentially valid line
     const bool is_end_of_invalid_line = (state_id == static_cast<StateT>(TT_INV) && is_delimiter);
 
     int32_t const emit_count =
-      is_end_of_invalid_line ? num_inv_tokens : (is_part_of_invalid_line && !is_delimiter ? 1 : 0);
+      is_end_of_invalid_line ? num_inv_tokens : (is_part_of_valid_line && !is_delimiter ? 1 : 0);
+    return emit_count;
+  }
+};
+
+struct TransduceTokenKeepLineEnd {
+  template <typename RelativeOffsetT, typename SymbolT>
+  constexpr CUDF_HOST_DEVICE SymbolT operator()(StateT const state_id,
+                                                SymbolGroupT const match_id,
+                                                RelativeOffsetT const relative_offset,
+                                                SymbolT const read_symbol) const
+  {
+    const bool is_end_of_invalid_line =
+      (state_id == static_cast<StateT>(TT_INV) &&
+       match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::DELIMITER));
+
+    const bool is_end_of_valid_line =
+      (state_id == static_cast<StateT>(TT_VLD) &&
+       match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::DELIMITER));
+
+    if (is_end_of_invalid_line) {
+      switch (relative_offset) {
+        case 0: return SymbolT{token_t::StructEnd, 0};
+        case 1: return SymbolT{token_t::StructBegin, 0};
+        case 2: return SymbolT{token_t::LineEnd, 0};
+        default: return SymbolT{token_t::LineEnd, 0};  // doesn't appear
+      }
+    } else if (is_end_of_valid_line) {
+      return SymbolT{token_t::LineEnd, 0};
+    } else {
+      return read_symbol;
+    }
+  }
+
+  template <typename SymbolT>
+  constexpr CUDF_HOST_DEVICE int32_t operator()(StateT const state_id,
+                                                SymbolGroupT const match_id,
+                                                SymbolT const read_symbol) const
+  {
+    // Number of tokens emitted on invalid lines
+    constexpr int32_t num_inv_tokens = 3;
+
+    const bool is_delimiter = match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::DELIMITER);
+
+    // If state is either invalid or we're entering an invalid state, we discard tokens
+    const bool is_part_of_valid_line =
+      (match_id != static_cast<SymbolGroupT>(dfa_symbol_group_id::ERROR) &&
+       state_id == static_cast<StateT>(TT_VLD));
+
+    // Indicates whether we transition from an invalid line to a potentially valid line
+    const bool is_end_of_invalid_line = (state_id == static_cast<StateT>(TT_INV) && is_delimiter);
+
+    int32_t const emit_count =
+      is_end_of_invalid_line ? num_inv_tokens : (is_part_of_valid_line || is_delimiter ? 1 : 0);
```

Comment:
Is it possible to merge the two translation functors? IIUC, there are two differences between `TransduceToken` and `TransduceTokenKeepLineEnd` 
1. If we are in valid state and read a delimiter token, then there is no output if `remove_line_end_token` is true, and the token is translated as-is otherwise.
2. If we are in invalid state and read a delimiter token, then the translation is `<SE, SB>` if `remove_line_end_token` is true, and `<LE, SE, SB>` otherwise.

If we have `remove_line_end_token` as a member of the functor, then we can have conditions like
```c++
    int32_t num_inv_tokens = remove_line_end_token ? 2 : 3;
    const bool is_delimiter = match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::DELIMITER);
    const bool valid_nops = (state_id == static_cast<StateT>(TT_VLD) && (remove_line_end_token ? 
                                                match_id != static_cast<SymbolGroupT>(dfa_symbol_group_id::OTHER_SYMBOLS) :
                                                match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::ERROR)));

    const bool invalid_nops = (state_id == static_cast<StateT>(TT_INV));
    const bool is_end_of_invalid_line = (state_id == static_cast<StateT>(TT_INV) && is_delimiter);

    int32_t const emit_count =
      is_end_of_invalid_line ? num_inv_tokens : (invalid_nops || valid_nops ? 0 : 1);
    return emit_count;
  }
``` 
I'm curious if something like this would work - we wouldn't need to add another kernel as other reviewers have mentioned.

Reply:
Thank you!! I've tweaked the implementation slightly for the same idea.